kv/kvserver: TestRequestsOnLaggingReplica failed

kv/kvserver.TestRequestsOnLaggingReplica [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20806344?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20806344?buildTab=artifacts#/) on master @ [ff95ef21d580eb15fa902299ae2e2dbaff8e0db7](https://github.com/cockroachdb/cockroach/commits/ff95ef21d580eb15fa902299ae2e2dbaff8e0db7):


```
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 2: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 2
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 2: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeat from store 1: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeat from store 1: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 2: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 2: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 2: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 2
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 2: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeat from store 1: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeat from store 1: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 2: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 2: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 2: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 2
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 2: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeat from store 1: to 3
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeat from store 1: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 2: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 2: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 2: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeat from store 3: to 2
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 3: to 2
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 2
    client_raft_helpers_test.go:557: dropping StoreLiveness msg MsgHeartbeatResp from store 1: to 3
    client_raft_helpers_test.go:559: allowing StoreLiveness msg MsgHeartbeatResp from store 2: to 3
    client_raft_test.go:1234: 
        	Error Trace:	pkg/kv/kvserver/client_raft_test.go:1234
        	            				pkg/testutils/subtest.go:23
        	Error:      	Should be false
        	Test:       	TestRequestsOnLaggingReplica/symmetric=false
--- FAIL: TestRequestsOnLaggingReplica/symmetric=false (27.99s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRequestsOnLaggingReplica.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57447

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -1086,6 +1086,9 @@ func TestRequestsOnLaggingReplica(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 
+	// Increase the verbosity of the test to help investigate failures.
+	testutils.SetVModule(t, "replica_range_lease=3,raft=4,replica_raft_quiesce=3,verify=2")
+
 	testutils.RunTrueAndFalse(t, "symmetric", func(t *testing.T, symmetric bool) {
 		st := cluster.MakeTestingClusterSettings()
 		kvserver.OverrideDefaultLeaseType(ctx, &st.SV, roachpb.LeaseLeader)
@@ -1219,6 +1222,21 @@ func TestRequestsOnLaggingReplica(t *testing.T) {
 		//
 		// NB: This relies on RejectLeaseOnLeaderUnknown being set to true.
 		log.KvExec.Infof(ctx, "test: sending request to partitioned replica")
+
+		if !symmetric {
+			// In asymmetric partition, wait for the partitioned node to learn about
+			// the new leader before continuing. We do this because we expect it to
+			// reply with a speculative lease below.
+			testutils.SucceedsSoon(t, func() error {
+				status := partitionedReplica.RaftStatus()
+				if status.Lead != 2 && status.Lead != 3 {
+					return errors.Errorf("partitioned replica doesn't know about new leader yet: lead=%d",
+						status.Lead)
+				}
+				return nil
+			})
+		}
+
 		getRequest := getArgs(key)
 		_, pErr := kv.SendWrapped(timeoutCtx, partitionedStoreSender, getRequest)
 		require.Error(t, pErr.GoError(), "unexpected success")

